src/size_hint.rs

```diff
@@ -1,149 +1,493 @@
+    upper: Upper,
+
+    out_of_fuel: bool,
+}
+
+#[derive(Clone, Copy, Eq, Hash, PartialEq)]
+enum Upper {
+    Bounded(usize),
+    /// The size is known to be unbounded (or overflow [`usize`]).
+    Unbounded,
+    /// The size is unknown either because a more specific size hint was not given,
+    /// or because computation of the size hint was cancelled before it completed.
+    /// This is functionally identical to `Unbounded` but prints differently.
+    ///
+    /// In these cases, the lower bound may also be lower than it should be.
+    Unknown,
+}
+
+impl SizeHint {
+    /// Equal to [`SizeHint::exactly(0)`][Self::exactly].
+    pub const ZERO: Self = Self {
+        lower: 0,
+        upper: Upper::Bounded(0),
+        out_of_fuel: false,
+    };
+
+    /// Indicates that a size hint is not available.
+    /// Returned by the default implementation of [`Arbitrary::size_hint()`].
+    ///
+    /// This is the most uninformative bound: zero to infinity bytes.
+    /// In addition, it prints differently from [`SizeHint::at_least(0)`][SizeHint::at_least],
+    /// to indicate that it is unknown rather than an unbounded collection.
+    pub const UNKNOWN: Self = Self {
+        lower: 0,
+        upper: Upper::Unknown,
+        out_of_fuel: false,
+    };
+
+    /// Indicates that the [`Context`] exhausted its `fuel`.
+    pub const OUT_OF_FUEL: Self = Self {
+        lower: 0,
+        upper: Upper::Unbounded,
+        out_of_fuel: true,
+    };
+
+    /// Creates a [`SizeHint`] for exactly `size` bytes.
+    #[must_use]
+    pub const fn exactly(size: usize) -> Self {
+        Self {
+            lower: size,
+            upper: Upper::Bounded(size),
+            out_of_fuel: false,
+        }
+    }
+
+    /// Creates a [`SizeHint`] for `size` bytes or more.
+    #[must_use]
+    pub const fn at_least(size: usize) -> Self {
+        Self {
+            lower: size,
+            upper: Upper::Unbounded,
+            out_of_fuel: false,
+        }
+    }
+
+    /// Returns the lower bound on bytes that will be consumed.
+    pub const fn lower_bound(self) -> usize {
+        self.lower
+    }
+
+    /// Returns the upper bound on bytes that will be consumed.
+    ///
+    /// Returns [`None`] if unlimited bytes could be consumed,
+    /// the number of bytes exceeds [`usize::MAX`],
+    /// or the bound is unknown.
+    pub const fn upper_bound(self) -> Option<usize> {
+        match self.upper {
+            Upper::Bounded(size) => Some(size),
+            Upper::Unbounded => None,
+            Upper::Unknown => None,
+        }
+    }
+
+    /// Take the sum of the `self` and `rhs` size hints.
+    ///
+    /// Use this when an [`Arbitrary::arbitrary()`] implementation is going to consume both
+    /// `self` bytes and `rhs` bytes, such as for the fields of a `struct`.
+    ///
+    /// This operation can also be written as the `+` operator.
+    #[inline]
+    pub const fn and(self, rhs: Self) -> Self {
```

Comment:
What do you think about moving all the and/and_all/or/or_all/repeat combinators to methods on `Context`? That way we retain the flexibility to make them consume fuel in the future, if that ever becomes necessary.